Add an index->step cache to the PolicyStepsRegistry

joegallo opened the pull request:
Related to #77466. Extracted and cleaned up from https://github.com/elastic/elasticsearch/pull/79692.

Adds a Index->Step cache (well, really Index->IndexMetadata+Step cache) that saves us from having to recalculate the step each time. It's keyed off the index, but if the IndexMetadata has changed, then the cache is bypassed and updated. Crucially, though, we also need to clear the cache in the event that the policies themselves change (because a policy change can result in the next step being different even if the IndexMetadata itself didn't change).

original-brownbear committed:
Cache for getStep

joegallo committed:
Add some comments

joegallo committed:
Move the delete logic from listener to applier

joegallo committed:
Clear out the cache entirely on master->non-master

joegallo committed:
Clear out the cache any time the policies change

joegallo commented:
Before this change (~20% of all samples in ILM, this `PolicyStepsRegistry.getSteps` highlighted with red rectangle):

![Screen Shot 2022-01-11 at 2 12 18 PM](https://user-images.githubusercontent.com/187034/149012771-e565c90b-4739-4d59-80fd-8316b51a75d3.png)

After this change (~5% of all samples in ILM):

![Screen Shot 2022-01-11 at 2 59 13 PM](https://user-images.githubusercontent.com/187034/149012789-eef0f536-a0f0-4021-aef7-b3ff1e60d0df.png)

joegallo committed:
Merge branch 'master' into ilm-policy-registry-steps-cache

original-brownbear commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistry.java:
```diff
@@ -305,7 +332,9 @@ public Step getStep(final IndexMetadata indexMetadata, final Step.StepKey stepKe
 
         // Return the step that matches the given stepKey or else null if we couldn't find it
-        return phaseSteps.stream().filter(step -> step.getKey().equals(stepKey)).findFirst().orElse(null);
+        final Step s = phaseSteps.stream().filter(step -> step.getKey().equals(stepKey)).findFirst().orElse(null);
```
Just a thought from looking at the profiling you posted: This might be happier as a normal loop :) The `phaseSteps` list is mostly (always?) very short and setting up a stream and doing all the dance around it doesn't even come close to outweighing the slightly faster iteration the stream provides. I'm good with this either way though, just a thought since this line was changed anyway :)

joegallo replied:
Ehhhhhhh, I don't think it's worth it. This reads pretty clearly as-is, and I don't think it's enough of a bottleneck anymore to warrant unrolling to a normal loop (due to this PR's optimization resulting in the code in question just being called a bajillion times less frequently).

elasticmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

original-brownbear commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistry.java:
```diff
@@ -64,6 +66,8 @@
     private final Map<String, Map<Step.StepKey, Step>> stepMap;
     private final NamedXContentRegistry xContentRegistry;
 
+    private final Map<Index, Tuple<IndexMetadata, Step>> cachedSteps = new ConcurrentHashMap<>();
```
Lets add a line or two of Javadoc explaining what this is for please. This is far from obvious for anybody not involved in this work.

joegallo replied:
How does 5f48e35b29ff8d9945bc5c14735409e0b84c549e sparkle with you?

original-brownbear replied:
>  IndexMetadata feels really heavyweight for this

We only use it for instance equality and clear it out of the cache if the index gets removed => memory wise it's free, instance equality comparison is cheap/free and this seems like the safest way of checking whether or not index metadata has changed.

We didn't just create a compound key here because we need/want the ability to quickly cleanup the cache if an index is removed without having to iterate a cache. Since we only care about the by index state this seemed fastest/easiest.

dakrone replied:
This has the side effect however, or any inconsequential change to the index metadata for the index (a dynamic field introduced, for example) invalidating the cache, whereas it doesn't actually need to be invalidated in that case?

original-brownbear replied:
I think that's ok. This is just supposed to help us with not running the expensive step computation on unchanged indices (which will always be the majority in almost any CS update). No need to super optimize this beyond that I'd say. Updating index metadata results in expensive operations anyway that even go as far as resulting in disk IO, the step calculation is probably trivial relatively speaking in all cases.

dakrone replied:
Alright, that's fine then, my preference is still to make the keyed part as small as possible, but this is okay for now I think.

original-brownbear reviewed (approved):
LGTM If we add some JavaDoc :)

joegallo committed:
Whitespace

joegallo committed:
Put these first fields in public constructor order

joegallo committed:
Add a comment that explains this cache

joegallo committed:
Merge branch 'master' into ilm-policy-registry-steps-cache

dakrone reviewed (changes requested):
I left a few comments, I think `IndexMetadata` is a bit big to hold on to for a cache like this, and I'd prefer to keep the map as small as possible, what do you think?

original-brownbear commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleService.java:
```diff
@@ -298,12 +298,19 @@ public void clusterChanged(ClusterChangedEvent event) {
             }
         }
 
+        for (Index index : event.indicesDeleted()) {
+            policyRegistry.delete(index);
+        }
```
huh come to think of it, can't we just return if `isMaster` is false, no need to load the ILM metadata below either? :)

dakrone replied:
That works also, we can return after the `cancelJob()` call above.

joegallo replied:
👍, 4954725add7628418152ab31ce684983266cfcf1

I didn't end up going with an early `return` but the logic is essentially the same as if I did.

dakrone commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistry.java:
```diff
@@ -267,6 +294,11 @@ public LifecyclePolicyMetadata read(StreamInput in, String key) {
     @Nullable
     public Step getStep(final IndexMetadata indexMetadata, final Step.StepKey stepKey) {
+        final Tuple<IndexMetadata, Step> cachedStep = cachedSteps.get(indexMetadata.getIndex());
+        if (cachedStep != null && cachedStep.v1() == indexMetadata && cachedStep.v2().getKey().equals(stepKey)) {
```
Can you add a comment about why instance equality is used here and okay rather than object equality?

joegallo replied:
👍, 68161e8a3b8168c5d9a64a55fe7bbbd30f0c8afe

joegallo commented:
@elasticmachine run elasticsearch-ci/packaging-tests-windows-sample

joegallo commented:
@elasticmachine run elasticsearch-ci/bwc

joegallo committed:
Merge branch 'master' into ilm-policy-registry-steps-cache

joegallo commented:
@elasticmachine run elasticsearch-ci/part-2

henningandersen commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleService.java:
```diff
@@ -298,12 +298,19 @@ public void clusterChanged(ClusterChangedEvent event) {
         }
 
+        for (Index index : event.indicesDeleted()) {
+            policyRegistry.delete(index);
```
I wonder if there is an (benign for sure) risk of a leak here: if a policy adds a step to the cache concurrently with it being deleted here, I think we may leave it there until next clear or master failover. Ideally we would retain only those that remain (thus curing this), but I can see how that is more expensive. A comment could be good though.

joegallo replied:
Sure thing, 19ecdd3f6860ffeb4fa097096f9f065ad4c3fcc5

henningandersen reviewed (commented):
Just a drive-by-comment.

joegallo commented:
Dropped the `7.17.0` label based on our conversation this earlier today.

original-brownbear commented:
@joegallo fyi we got the green-light to  backport this to 7.17 after all!

joegallo committed:
Merge branch 'master' into ilm-policy-registry-steps-cache

joegallo committed:
Update docs/changelog/82316.yaml

elasticsearchmachine commented:
Hi @joegallo, I've created a changelog YAML for you.

joegallo committed:
Add some explanatory comments

joegallo committed:
Surround this logic in an isMaster check

Promoting the isMaster check around triggerPolicies up to include both
these blocks.

joegallo commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistry.java:
```diff
@@ -155,6 +166,22 @@ public LifecyclePolicyMetadata read(StreamInput in, String key) {
         }
     }
 
+    public void delete(Index deleted) {
```
👍, 68161e8a3b8168c5d9a64a55fe7bbbd30f0c8afe

joegallo committed:
Add a note about a race condition

joegallo committed:
Merge branch 'master' into ilm-policy-registry-steps-cache

dakrone reviewed (approved):
LGTM
